politeiad/backendv2/tstorebe/plugins/ticketvote/hooks.go

```diff
@@ -0,0 +1,370 @@
+	switch {
+	case vm.LinkBy == 0 && vm.LinkTo == "":
+		// Vote metadata is empty
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeVoteMetadataInvalid),
+			ErrorContext: "metadata is empty",
+		}
+
+	case vm.LinkBy != 0 && vm.LinkTo != "":
+		// LinkBy and LinkTo cannot both be set
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeVoteMetadataInvalid),
+			ErrorContext: "cannot set both linkby and linkto",
+		}
+
+	case vm.LinkBy != 0:
+		err := p.linkByVerify(vm.LinkBy)
+		if err != nil {
+			return err
+		}
+
+	case vm.LinkTo != "":
+		err := p.linkToVerify(vm.LinkTo)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *ticketVotePlugin) hookNewRecordPre(payload string) error {
+	var nr plugins.HookNewRecordPre
+	err := json.Unmarshal([]byte(payload), &nr)
+	if err != nil {
+		return err
+	}
+
+	// Verify the vote metadata if the record contains one
+	vm, err := voteMetadataDecode(nr.Files)
+	if err != nil {
+		return err
+	}
+	if vm != nil {
+		err = p.voteMetadataVerify(*vm)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *ticketVotePlugin) hookEditRecordPre(payload string) error {
+	var er plugins.HookEditRecord
+	err := json.Unmarshal([]byte(payload), &er)
+	if err != nil {
+		return err
+	}
+
+	// The LinkTo field is not allowed to change once the record has
+	// become public. If this is a vetted record, verify that any
+	// previously set LinkTo has not changed.
+	if er.Current.RecordMetadata.State == backend.StateVetted {
+		var (
+			oldLinkTo string
+			newLinkTo string
+		)
+		vm, err := voteMetadataDecode(er.Current.Files)
+		if err != nil {
+			return err
+		}
+		if vm != nil {
+			oldLinkTo = vm.LinkTo
+		}
+		vm, err = voteMetadataDecode(er.Files)
+		if err != nil {
+			return err
+		}
+		if vm != nil {
+			newLinkTo = vm.LinkTo
+		}
+		if newLinkTo != oldLinkTo {
+			e := fmt.Sprintf("linkto cannot change on vetted record: "+
+				"got '%v', want '%v'", newLinkTo, oldLinkTo)
+			return backend.PluginError{
+				PluginID:     ticketvote.PluginID,
+				ErrorCode:    uint32(ticketvote.ErrorCodeLinkToInvalid),
+				ErrorContext: e,
+			}
+		}
+	}
+
+	// Verify LinkBy if one was included. The VoteMetadata is optional
+	// so the record may not contain one.
+	vm, err := voteMetadataDecode(er.Files)
+	if err != nil {
+		return err
+	}
+	if vm != nil {
+		err = p.linkByVerify(vm.LinkBy)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *ticketVotePlugin) hookSetRecordStatusPre(payload string) error {
+	var srs plugins.HookSetRecordStatus
+	err := json.Unmarshal([]byte(payload), &srs)
+	if err != nil {
+		return err
+	}
+
+	// Check if the LinkTo has been set
+	vm, err := voteMetadataDecode(srs.Current.Files)
+	if err != nil {
+		return err
+	}
+	if vm != nil && vm.LinkTo != "" {
```

Comment:
I see this if all over the code. Should this maybe become a function that returns a bool making all invocations more readable?

Reply:
This code has been refactored to use functions that return if the vote metadata is not present, instead of having `vm == nil` everywhere.

https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/plugins/ticketvote/hooks.go